Add custom error handling (#64)
cloudflare/kv-asset-handler: README.md

# @cloudflare/kv-asset-handler
  * [Installation](#installation)
  * [Usage](#usage)
  * [`getAssetFromKV`](#-getassetfromkv-)
      - [Example](#example)
    + [Return](#return)
    + [Optional Arguments](#optional-arguments)
      - [`mapRequestToAsset`](#-maprequesttoasset-)
      - [Example](#example-1)
      - [`cacheControl`](#-cachecontrol-)
        * [`browserTTL`](#-browserttl-)
        * [`edgeTTL`](#-edgettl-)
        * [`bypassCache`](#-bypasscache-)
      - [`ASSET_NAMESPACE`](#-asset-namespace-)
      - [`ASSET_MANIFEST` (optional)](#-asset-manifest---optional-)
- [Helper functions](#helper-functions)
  * [`mapRequestToAsset`](#-maprequesttoasset--1)

>   * [`serveSinglePageApp`](#-servesinglepageapp-)

Reviewer: nice TOC 👍

### `getAssetFromKV`

> `getAssetFromKV` is a function that takes a `FetchEvent` object and returns a `Response` object if the request matches an asset in KV, otherwise it will throw a `KVError`.

Reviewer: where do we intend to document the different KVErrors one could get back?

> `getAssetFromKV` is a function that takes a `FetchEvent` object and returns a `Response` object if the request matches an asset in KV, otherwise it will throw a `KVError`.

Reviewer: where do we intend to document the possible KV errors this might return?

Author: Good idea done

> `getAssetFromKV` is a function that takes a `FetchEvent` object and returns a `Response` object if the request matches an asset in KV, otherwise it will throw an `KVError`.

Reviewer 2: ```suggestion
`getAssetFromKV` is a function that takes a `FetchEvent` object and returns a `Response` object if the request matches an asset in KV, otherwise it will throw a `KVError`.
```

Reviewer: i think it is worth calling out the different kv error types, as well.

This package was designed to work with [Worker Sites](https://workers.cloudflare.com/sites).

## `getAssetFromKV`

getAssetFromKV(FetchEvent) => Promise<Response>

`getAssetFromKV` is an async function that takes a `FetchEvent` object and returns a `Response` object if the request matches an asset in KV, otherwise it will throw a `KVError`.

#### Example

This example checks for the existence of a value in KV, and returns it if it's there, and returns a 404 if it is not. It also serves index.html from `/`.

### Return

> `getAssetFromKV` returns a `Promise<Response>` with `Response` being the body of the asset requested.

Reviewer: ```suggestion
`getAssetFromKV` returns a `Promise<Response>` where the `Response` body contains the asset requested. It will reject with one of the KVError variants listed below.
```

> Known errors to be thrown are:

Reviewer: ```suggestion
KVError variants include:
```

Author: that's not accurate unknown errors can be thrown

- MethodNotAllowedError
- NotFoundError
- InternalError
  
```js
import { getAssetFromKV } from '@cloudflare/kv-asset-handler'

addEventListener('fetch', event => {
  event.respondWith(handleEvent(event))
})

async function handleEvent(event) {
  if (event.request.url.includes('/docs')) {
    try {
      return await getAssetFromKV(event)
    } catch (e) {
        switch (typeof resp) {
          case NotFoundError: 
            //..
          case MethodNotAllowedError:

>             // ...

Reviewer: i like this.

This package was designed to work with [Worker Sites](https://workers.cloudflare.com/sites).

## `getAssetFromKV`

> type: function(FetchEvent) => Promise<Response>

Reviewer 3: why not:
```suggestion
getAssetFromKV(FetchEvent) => Promise<Response>
```

    try {
      return await getAssetFromKV(event, { mapRequestToAsset: customKeyModifier })
    } catch (e) {
      if (resp instanceof KVError) {
        switch (e.code) {
          case 404:

>             return notFoundResponse

Reviewer 2: ```suggestion
            return new Response(`${event.request.url} was not found`
```

Reviewer 3: Shouldn't this pull from 404.html in KV?

Author: This is just the README so i wanted to keep it as simple as possible

>             break

Reviewer 2: ```suggestion

```

          case 405:

>             return methodNotAllowedResponse

Reviewer 2: ```suggestion
            return new Response(`method ${event.request.method} not allowed`}
```

Author: ingMeh I could go either way on this but I think since it's the README, keep this line short and sweet prevents overwhelming the reader with an already complex example

Reviewer: i do appreciate the want for simplicity; however it might make it seem like `methodNotAllowedResponse` is a pre-defined thing available to the user.

Author: This is just the README it's not meant to be functional, even copy pasteable code but a guide to seeing what the function does on a high level

Reviewer: people copy paste code out of readmes

Author: I disagree. Maybe we can address this is a separate issue? I see other areas to improve the readme as well that I don't want to block this

Reviewer: there are a lot of ways to improve the readme, but as it is right now, all the code examples in this readme are functioning code.

Author: The current snippet didn't actually function. I removed the specifics on handling the errors and it now functions

>             break

Reviewer 2: ```suggestion

```

          default:
            return new Response(resp.message, { status: resp.code })
        }
      }

>       return new Response(`Finding "${customKeyModifier(event.request.url)}" threw an error`)

Reviewer 2: ```suggestion
      return new Response("An unexpected error occurred", {status: 500})
```

  event.respondWith(handleEvent(event))
})

async function handleEvent(event) {
  if (event.request.url.includes('/docs')) {
    try {
      return await getAssetFromKV(event)
    } catch (e) {
      if (resp instanceof KVError) {

>         switch (e.status) {

Reviewer 3: Ooh! I thought we were going to match over error types (and I honestly did prefer that approach!) I feel like matching on types like NotFoundError would look nice in the docs, too, especially because we describe the error types right below this example.

          case 404:
            // ...
        }
      }
      return new Response("An unexpected error occurred", { status: 500 })
    }
  } else return fetch(event.request)
}
```

> ### Return

Reviewer 3: I feel like hoisting these error return types above the example would give a lot of valuable context.

  event.respondWith(handleEvent(event))
})

async function handleEvent(event) {
  if (event.request.url.includes('/docs')) {
    try {

>       return await getAssetFromKV(event)

Author: Remove modifying `mapRequestToAsset` to the first example since it's illustrated below
